src/fits.cc

```diff
@@ -912,6 +936,131 @@ void Fits::writeImageImpl(T const *data, int nElements) {
+    // cfitsio handle it itself.
+    // In all other cases, we will convert the image to use the appropriate scale and zero
+    // (because we want to fuzz the numbers in the quantisation), so we don't want cfitsio
+    // rescaling.
+    if (!std::is_same<T, std::uint64_t>::value) {
+        fits_set_bscale(fits, 1.0, 0.0, &status);
+        if (behavior & AUTO_CHECK) {
+            LSST_FITS_CHECK_STATUS(*this, "Setting bscale,bzero");
+        }
+    }
+
+    // Write the pixels
+    int const fitsType = scale.bitpix == 0 ? FitsType<T>::CONSTANT : fitsTypeForBitpix(scale.bitpix);
+    fits_write_img(fits, fitsType, 1, pixels->getNum(), pixels->getData(), &status);
+    if (behavior & AUTO_CHECK) {
+        LSST_FITS_CHECK_STATUS(*this, "Writing image");
+    }
+
+    if (!std::is_same<T, std::int64_t>::value && !std::is_same<T, std::uint64_t>::value &&
+        std::isfinite(scale.bzero) && std::isfinite(scale.bscale) && (scale.bscale != 0.0)) {
+        if (std::numeric_limits<T>::is_integer) {
+            if (scale.bzero != 0.0) {
+                fits_write_key_lng(fits, "BZERO", static_cast<long>(scale.bzero),
+                                   "Scaling: MEMORY = BZERO + BSCALE * DISK", &status);
+            }
+            if (scale.bscale != 1.0) {
+                fits_write_key_lng(fits, "BSCALE", static_cast<long>(scale.bscale),
+                                   "Scaling: MEMORY = BZERO + BSCALE * DISK", &status);
+            }
+        } else {
+            fits_write_key_dbl(fits, "BZERO", scale.bzero, 12,
+                               "Scaling: MEMORY = BZERO + BSCALE * DISK", &status);
+            fits_write_key_dbl(fits, "BSCALE", scale.bscale, 12,
+                               "Scaling: MEMORY = BZERO + BSCALE * DISK", &status);
+        }
+        if (behavior & AUTO_CHECK) {
+            LSST_FITS_CHECK_STATUS(*this, "Writing BSCALE,BZERO");
+        }
+    }
+
+    if (scale.bitpix > 0) {
+        fits_write_key_lng(fits, "BLANK", scale.blank, "Value for undefined pixels", &status);
+        fits_write_key_lng(fits, "ZBLANK", scale.blank, "Value for undefined pixels", &status);
+        if (!std::numeric_limits<T>::is_integer) {
+            fits_write_key_lng(fits, "ZDITHER0", options.scaling.seed, "Dithering seed", &status);
+            fits_write_key_str(fits, "ZQUANTIZ", "SUBTRACTIVE_DITHER_1", "Dithering algorithm", &status);
+        }
+        if (behavior & AUTO_CHECK) {
+            LSST_FITS_CHECK_STATUS(*this, "Writing [Z]BLANK");
+        }
+    }
+
+    // cfitsio says this is deprecated, but Pan-STARRS found that it was sometimes necessary, writing:
+    // "This forces a re-scan of the header to ensure everything's kosher.
+    // Without this, compressed HDUs have been written out with PCOUNT=0 and TFORM1 not correctly set."
+    fits_set_hdustruc(fits, &status);
+    if (behavior & AUTO_CHECK) {
+        LSST_FITS_CHECK_STATUS(*this, "Finalizing header");
+    }
```

Comment:
That's what needs to be done to write a compressed image. I could strip parts out into other functions, but they wouldn't be used for anything else.

Some header keywords are written after the pixels because we don't want cfitsio to know about them we were writing the pixels (because we don't want it using them to modify the pixels, and we don't want it overwriting the values we chose). I added a comment explaining this.